rust/foxglove/src/websocket/protocol/server.rs

```diff
@@ -178,30 +181,46 @@ pub fn server_info(
     .to_string()
 }
 
+// A `schema` in the channel is optional except for message_encodings which require a schema.
+// Currently, Foxglove supports schemaless JSON messages.
 // https://github.com/foxglove/ws-protocol/blob/main/docs/spec.md#advertise
-// Caller must check that the channel has a schema, otherwise this will panic.
 pub fn advertisement(channel: &Channel) -> Result<String, FoxgloveError> {
-    let schema = &channel
-        .schema
-        .as_ref()
-        .ok_or_else(|| FoxgloveError::SchemaRequired)?;
+    let schema = channel.schema.as_ref();
 
-    let schema_data = if super::is_known_binary_schema_encoding(&schema.encoding) {
-        BASE64_STANDARD.encode(&schema.data)
-    } else {
-        String::from_utf8(schema.data.to_vec()).map_err(|e| FoxgloveError::Unspecified(e.into()))?
-    };
+    if schema.is_none() && REQUIRING_SCHEMA.contains(&channel.message_encoding.as_str()) {
+        return Err(FoxgloveError::SchemaRequired);
+    }
 
-    Ok(json!({
-        "op": "advertise",
-        "channels": [Advertisement{
+    let advertisement = if let Some(schema) = schema {
+        let schema_data = if super::is_known_binary_schema_encoding(&schema.encoding) {
```

Comment:
nit: it now seems odd that `is_known_binary_schema_encoding` is a function while `REQUIRING_SCHEMA` is a const array -- feel free to change mine if you prefer the array style (I don't think I have a strong preference)

Reply:
Thanks, I didn't notice this after merging that change from main. I don't have a strong preference either; I'll rewrite mine with a function.